### ISSUE TITLE
KeyError when adding PII field to a child table

### ISSUE THREAD

**Issue by halvorot** [sdv-dev/SDV#846]
### Environment Details

Please indicate the following details about the environment in which you found the bug:

* SDV version: 0.15.0
* Python version: 3.9
* Operating System: Windows

### Error Description

I am not sure if it is a bug or I am doing something wrong here but I have a database with three tables and training a HMA1 model on it. See Image and code below for the metadata.

The problem occurs when I try to add a field as PII in a specific one of the tables (forerettighetsbevis), it works fine with the other two tables (trafikant, forerettighet).

Do you have any idea what could cause this error when adding PII fields? Please let me know if you need any more information about the problem.

The log/traceback is:
```
09:23:18 INFO: Modeling trafikant
09:23:18 INFO: Computing extensions for table trafikant
09:23:18 INFO: Modeling forerettighetsbevis
09:23:18 INFO: Computing extensions for table forerettighetsbevis
09:23:18 INFO: Fitting GaussianCopula for table forerettighetsbevis; shape: (3, 14)
09:23:18 INFO: Fitting table forerettighetsbevis metadata
09:23:18 INFO: Anonymizing table forerettighetsbevis
09:23:18 INFO: Fitting constraints for table forerettighetsbevis
09:23:18 INFO: Fitting HyperTransformer for table forerettighetsbevis
Traceback (most recent call last):
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\pandas\core\indexes\base.py", line 3621, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas\_libs\index.pyx", line 136, in pandas._libs.index.IndexEngine.get_loc
  File "pandas\_libs\index.pyx", line 163, in pandas._libs.index.IndexEngine.get_loc
  File "pandas\_libs\hashtable_class_helper.pxi", line 5198, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas\_libs\hashtable_class_helper.pxi", line 5206, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: '__forerettighetsbevis__TRAFIKANT_ID__num_rows'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\hteigen\Documents\GIT\sdv-app\syntheticlib\run.py", line 113, in <module>
    model  = load_train(args.schema)
  File "C:\Users\hteigen\Documents\GIT\sdv-app\syntheticlib\run.py", line 83, in load_train
    model.train_model(tables_dict, metadata)
  File "C:\Users\hteigen\Documents\GIT\sdv-app\syntheticlib\syntheticlib\Model.py", line 25, in train_model
    self.model.fit(tables=raw_data_dict)
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\sdv\relational\base.py", line 63, in fit
    self._fit(tables)
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\sdv\relational\hma.py", line 281, in _fit
    self._model_table(table_name, tables)
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\sdv\relational\hma.py", line 244, in _model_table
    table = self._extend_table(table, tables, table_name)
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\sdv\relational\hma.py", line 164, in _extend_table
    table[num_rows_key].fillna(0, inplace=True)
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\pandas\core\frame.py", line 3505, in __getitem__
    indexer = self.columns.get_loc(key)
  File "C:\Users\hteigen\Anaconda3\envs\syntheticlib\lib\site-packages\pandas\core\indexes\base.py", line 3623, in get_loc
    raise KeyError(key) from err
KeyError: '__forerettighetsbevis__TRAFIKANT_ID__num_rows'
```

![image](https://user-images.githubusercontent.com/36247551/174247629-513c31d4-f76e-4121-8b08-c43d8dd73048.png)
```json
{
    "tables": {
        "trafikant": {
            "fields": {
                "ID": {
                    "type": "id",
                    "subtype": "integer"
                },
                "KUNDE_ID": {
                    "type": "categorical"
                },
                "ENDRET_AV": {
                    "type": "categorical"
                },
                "ENDRET_DATO": {
                    "type": "datetime"
                },
                "FODELAND": {
                    "type": "categorical"
                },
                "FODSELSDATO": {
                    "type": "datetime"
                },
                "KJONN": {
                    "type": "categorical"
                },
                "DODS_DATO": {
                    "type": "datetime"
                },
                "FORERKORT_FORSTE_GANG": {
                    "type": "datetime"
                }
            },
            "primary_key": "ID"
        },
        "forerettighet": {
            "fields": {
                "ID": {
                    "type": "id",
                    "subtype": "integer"
                },
                "GJELDER_FOM": {
                    "type": "datetime"
                },
                "GJELDER_TOM": {
                    "type": "datetime"
                },
                "KLADD": {
                    "type": "categorical"
                },
                "TRAFIKANT_ID": {
                    "type": "id",
                    "subtype": "integer",
                    "ref": {
                        "table": "trafikant",
                        "field": "ID"
                    }
                },
                "KLASSE_ID": {
                    "type": "categorical"
                }
            },
            "primary_key": "ID"
        },
        "forerettighetsbevis": {
            "fields": {
                "ID": {
                    "type": "id",
                    "subtype": "integer"
                },
                "BEVISTYPE": {
                    "type": "categorical"
                },
                "BEVISNUMMER": {
                    "type": "categorical"
                },
                "FORSTEGANGSUTSTEDELSE": {
                    "type": "categorical"
                },
                "UTSTEDER_NAVN": {
                    "type": "categorical"
                },
                "DUPLIKAT": {
                    "type": "categorical"
                },
                "FNR": {
                    "type": "categorical"
                },
                "ETTERNAVN": {
                    "type": "categorical",
                    "pii": true,
                    "pii_category": "last_name"
                },
                "MALFORM": {
                    "type": "categorical"
                },
                "POSTNR": {
                    "type": "categorical"
                },
                "POSTSTED": {
                    "type": "categorical"
                },
                "FORNAVN_MELLOMNAVN": {
                    "type": "categorical"
                },
                "UTSTEDER_ID": {
                    "type": "categorical"
                },
                "GJELDENDE_BEVIS": {
                    "type": "categorical"
                },
                "TRAFIKANT_ID": {
                    "type": "id",
                    "subtype": "integer",
                    "ref": {
                        "table": "trafikant",
                        "field": "ID"
                    }
                },
                "UTSTEDT_DATO": {
                    "type": "datetime"
                }
            },
            "primary_key": "ID"
        }
    }
}
```

**Comment by npatki** [comment]
Hi @halvorot, thanks for filing this. Your metadata file and schema look correct to me.

I can replicate this issue on a similar schema. It seems like there is a KeyError from adding a PII field whenever the PII field is on a child table _and_ that child table is not the first child table. Let me update this issue title with the specifics. 

We'll keep this bug open to ask any new questions and track updates. If you discover anything more about when this does/does not work, feel free to respond.

**Comment by npatki** [comment]
Actually I am also seeing this issue if the PII field is put on the first child (`forerettighet`, column `KLADD`).

**Comment by montasIET** [comment]
I'm having the same problem. Any news about this?

**Comment by npatki** [comment]
Hi @montasIET, we are aware of the issue. A future release of SDV will include a general code refactor that should solve it.

Temporary workaround for now (I recognize it's not ideal):

1. Drop the PII column from the table. Update the metadata to exclude the column.
```python
child_table_no_pii = child_table.drop(['pii_column_name'], axis=1)
```

2. Fit the SDV on the data without the PII column and sample
3. Re-create the PII column using `Faker` directly. (This is what SDV uses under-the-hood.)
```python
from faker import Faker

fake = Faker()
num_rows = 100

# replace 'lexify' with whatever your PII type is 
# eg 'zipcode', 'street_address', 'phone_number', etc.
new_pii_values = [fake.lexify() for i in range(num_rows)]

synthetic_child_table['pii_column_name'] = new_pii_values
```

**Comment by montasIET** [comment]
@npatki I've done a little bit of reengineering and I still have the same problem (KeyError: table name), but this time I don't have PII field in children. I only have a main table with a primary key and two tables with the reference key, and in both of them there are not PII field. Is there a way to solve it?

**Comment by montasIET** [comment]
Hi @npatki have you inspected this solution? Can it be a good workaround?

**Comment by npatki** [comment]
Hi @halvorot and @montasIET -- great news! This bug has now been resolved in our new, [**SDV 1.0 (Beta!)](https://docs.sdv.dev/sdv) release. Check it out and let us know if you're still seeing any problems!

For more resources see:
- New [HMASynthesizer docs](https://docs.sdv.dev/sdv/multi-table-data/modeling/hmasynthesizer)
- [Demo](https://colab.research.google.com/drive/1h_xwzHXrif-xEOC39amYC6xfNPu5hDA1?usp=sharing) for synthesizing multiple tables
